packages/cli/src/metadataGeneration/typeResolver.ts

```diff
@@ -152,143 +156,233 @@ export class TypeResolver {
         );
       };
-      const properties: Tsoa.Property[] = type
-        .getProperties()
-        // Ignore methods, getter, setter and @ignored props
-        .filter(property => isIgnored(property) === false)
-        // Transform to property
-        .map(property => {
-          const propertyType = typeChecker.getTypeOfSymbolAtLocation(property, this.typeNode);
-          const declaration = getDeclaration(property) as ts.PropertySignature | ts.PropertyDeclaration | ts.ParameterDeclaration | undefined;
-
-          if (declaration && ts.isPropertySignature(declaration)) {
-            return { ...this.propertyFromSignature(declaration, mappedTypeNode.questionToken), name: property.getName() };
-          } else if (declaration && (ts.isPropertyDeclaration(declaration) || ts.isParameter(declaration))) {
-            return { ...this.propertyFromDeclaration(declaration, mappedTypeNode.questionToken), name: property.getName() };
-          }
-
-          // Resolve default value, required and typeNode
-          let required = false;
-          const typeNode = this.current.typeChecker.typeToTypeNode(propertyType, undefined, ts.NodeBuilderFlags.NoTruncation)!;
-          if (mappedTypeNode.questionToken && mappedTypeNode.questionToken.kind === ts.SyntaxKind.MinusToken) {
-            required = true;
-          } else if (mappedTypeNode.questionToken && mappedTypeNode.questionToken.kind === ts.SyntaxKind.QuestionToken) {
-            required = false;
-          }
 
-          // Push property
+      const calcMappedType = (type: ts.Type): Tsoa.Type => {
+        if (type.flags & ts.TypeFlags.Union) {
+          //Intersections are not interesting somehow...
+          const types = (type as ts.UnionType).types;
+          const resolvedTypes = types.map(calcMappedType);
           return {
-            name: property.getName(),
-            required,
-            // Mapped types with any amount of indirection (template strings, unions, Exclude<>, etc.)
-            // don't provide an underlying declaration in the AST, thus we cannot know if the original
-            // property is deprecated. This matches intellisense's behavior in vscode.
-            deprecated: false,
-            type: new TypeResolver(typeNode, this.current, this.typeNode, this.context, this.referencer).resolve(),
-            validators: {},
+            dataType: 'union',
+            types: resolvedTypes,
           };
-        });
-
-      const objectLiteral: Tsoa.NestedObjectLiteralType = {
-        dataType: 'nestedObjectLiteral',
-        properties,
+        } else if (type.flags & ts.TypeFlags.Undefined) {
+          return {
+            dataType: 'undefined',
+          };
+        } else if (type.flags & ts.TypeFlags.Null) {
+          return {
+            dataType: 'enum',
+            enums: [null],
+          };
```

Comment:
undefined & null types are interesting if strictNullChecks flag is true in ts compilerOptions